PyCaret - LORIS Tutorial (#5539)
galaxyproject/training-material: topics/statistics/tutorials/loris_model/tutorial.md

---
layout: tutorial_hands_on
level: Intermediate

> title: Building the LORIS LLR6 PanCancer Model Using Galaxy-PyCaret Tool

Reviewer: ```suggestion
title: Building the LORIS LLR6 PanCancer Model Using PyCaret
```

zenodo_link: https://zenodo.org/records/13885908
questions:
- How can I reproduce the LORIS LLR6 model published by Chang et al., 2024?
- Which tools can I use in Galaxy to obtain a logistic regression model?

> - How can I evaluate the model to confirm its performance?

Reviewer 2: ```suggestion
- How can the model be evaluated to confirm its performance?
```

objectives:
- Use a large dataset of immune checkpoint blockade (ICB)-treated and non-ICB-treated patients across 18 solid tumor types, encompassing a wide range of clinical, pathologic and genomic features to build a Machine Learning Model.

> - Build a Machine Learing model using PyCaret tool available in Galaxy.

Reviewer: ```suggestion
- Build a Machine Learning model using PyCaret in Galaxy.
```

> - Evaluate the model for robustness by comparing it with the original LORIS LLR6 model published by Chang et al., 2024.

Reviewer 2: The LORIS LLR6 model should be discussed in the tutorial. Details such as model architecture, features etc should be mentioned. It would improve the tutorial.

Author: Thank you. I have included more information regarding the LLR6 model.

> <agenda-title></agenda-title>
>
> In this tutorial, we will cover:
>
> 1. TOC
> {:toc}
>
{: .agenda}

> <comment-title>Background</comment-title>
>
> The [LORIS dataset](https://github.com/rootchang/LORIS/blob/main/02.Input/AllData.xlsx) contains clinical, pathologic, and genomic features from a diverse cohort of patients, providing a foundation for robust analysis and model development. The raw dataset (xlsx) includes 10 distinct cohorts:
>
> Chowell_train, 2) Chowell_test, 3) MSK1, 4) MSK2, 5) Kato_panCancer, 6) Shim_NSCLC, 7) Vanguri_NSCLC, 8) Ravi_NSCLC, 9) Pradat_panCancer, 10) MSK_nonICB.
>
> To keep this tutorial concise, we will focus on the Chowell_train cohort for model training and the Chowell_test cohort for testing the PanCancer model.
>
{:  .comment}

# Dataset composition to train the model 
Before we begin the hands-on session, here’s a brief explanation of the features we’ll be using. These features were selected based on the findings of Chang et al. (2024), who identified them as the most important for training the model.

## TMB (Tumor mutation burden)

> Tumor mutation burden (TMB) is defined as the total number of somatic mutations within a specified region of the tumor genome. It has been recognized as a biomarker for predicting the efficacy of immune checkpoint blockade (ICB) in solid tumors. The FDA has approved a threshold of 10 mutations per megabase (mut/Mb) as a biomarker for response to ICB treatment.

Reviewer 2: ```suggestion
Tumor mutation burden (TMB) is defined as the total number of somatic mutations within a specified region of the tumor genome. It has been recognized as a biomarker for predicting the efficacy of immune checkpoint blockade (ICB) in solid tumors. The [U.S. Food and Drug Administration](https://www.fda.gov/) (FDA) has approved a threshold of 10 mutations per megabase (mut/Mb) as a biomarker for response to ICB treatment.
```

## Response
The Response feature is a categorical target variable indicating whether patients benefited from immune checkpoint blockade (ICB) therapy, classified as 0 (no benefit) or 1 (benefit). The model is trained using all previous features to predict patient outcomes when ICB is chosen as the treatment.

# Prepare environment and get the data 
> <comment-title>Preprocessing the raw data</comment-title>
>
> The raw data published by ({% cite Chang2024 %}) can be found here: 
> [LORIS raw dataset](https://github.com/rootchang/LORIS/blob/main/02.Input/AllData.xlsx)
> 
> We preprocessed the raw data using a Python script to:
> 1) Extract the `Chowell_train` and `Chowell_test` tab from the excel file.
> 2) Select the 7 features (`TMB`, `Systemic Therapy History`, `Albumin`, `Cancer Type`, `NLR`, `Age`, and `Response`) important for building the model.
> 3) Truncate the values for `Age`, `NLR`, and `TMB`.
> 4) Encode `Cancer Type` using one-hot encoding. 
> 5) Save the dataset as a .tsv file.
>

> > A jupyter notebook for preprocessing can be found at Dockstore: [LORIS_preprocessing](https://dockstore.org/notebooks/github.com/paulocilasjr/pycaret-use-case/preprocessing:main?tab=info)

Reviewer: ```suggestion
> A Jupyter notebook for preprocessing can be found at Dockstore: [LORIS_preprocessing](https://dockstore.org/notebooks/github.com/paulocilasjr/pycaret-use-case/preprocessing:main?tab=info)
```

>
{:  .comment}

> <hands-on-title> Environment and Data Upload </hands-on-title>
>
> 1. Create a new history for this tutorial. If you are not inspired, you can name it *LORIS model classifier*.
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import the files from Zenodo or from the shared data library
>
>    ```
>    https://zenodo.org/records/13885908/files/Chowell_train_Response.tsv
>    https://zenodo.org/records/13885908/files/Chowell_test_Response.tsv
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
> 3. Check that the data format assigned for the file is **tsv**.
>    If it is not, follow the Changing the datatype tip.
>
>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="datatypes" %}
>
> 5. Add a tag (`LORIS model dataset`) to the dataset corresponding to `Chowell_train_Response.tsv` and `Chowell_test_Response.tsv`
>    This is important to trace back on what dataset the model was built on.
>
>    {% snippet faqs/galaxy/datasets_add_tag.md %}
>
{: .hands_on}

# Using PyCaret Model Comparison Tool
> <hands-on-title> Task description </hands-on-title>
>

> > 1. {% tool [PyCaret Model Comparison](.bx.psu.edu/repos/paulo_lyra_jr/pycaret_model_comparison/PyCaret_Model_Comparison/2024.3.3.2+0) %} with the following parameters:

Reviewer: the tool-id looks broken

Author: ![Screenshot 2024-12-02 at 4 06 49 PM](https://github.com/user-attachments/assets/326e8488-5b0e-4882-8d00-353782db08a2)
This one is odd. I changed because of lint test. See here the change.

Reviewer: What was the old linting error?

Reviewer: This link is def wrong, can you please change it back?

Author: I looked other tutorial to follow same pattern. Found this one: [regression_machinelearning](https://raw.githubusercontent.com/paulocilasjr/training-material/refs/heads/main/topics/statistics/tutorials/regression_machinelearning/tutorial.md).

it will satisfy the test and wont be broken.

>    - {% icon param-file %} *"Input Dataset (CSV or TSV)"*: `Chowell_train_Response.tsv`
>    - {% icon param-file %} *"Test Dataset (CSV or TSV)"*: `Chowell_test_Response.tsv`
>    - {% icon param-file %} *"Select the target column"*: `C22: Response`
>    - {% icon param-file %} *"Task"*: `Classification`
> Run the tool 
{: .hands_on}

# Tool output files
After training and testing your model, you should see two new files in your history list:

- PyCaret Model Comparison Best Model: The PyCaret model pickle file. This file allows the model to be reused without requiring retraining, ensuring consistent predictions.

- PyCaret Model Report: This file contains all the plots for the models trained, along with the best model selected.

> For the purposes of this tutorial, we will focus on the PyCaret Model Report.

Reviewer: ```suggestion
For this tutorial, we will focus on the PyCaret Model Report.
```

> 2) Select the 7 features (`TMB`, `Systemic Therapy History`, `Albumin`, `Cancer Type`, `NLR`, `Age`, and `Response`) important for building the model.
> 3) Truncate the values for `Age`, `NLR`, and `TMB`.
> 4) Encode `Cancer Type` using one-hot encoding. 
> 5) Save the dataset as a .tsv file.
>
> A Jupyter notebook for preprocessing can be found at Dockstore: [LORIS_preprocessing](https://dockstore.org/notebooks/github.com/paulocilasjr/pycaret-use-case/preprocessing:main?tab=info)
>
{:  .comment}

> <hands-on-title> Environment and Data Upload </hands-on-title>
>
> 1. Create a new history for this tutorial. If you are not inspired, you can name it *LORIS model classifier*.
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import the files from Zenodo or from the shared data library
>
>    ```
>    https://zenodo.org/records/13885908/files/Chowell_train_Response.tsv
>    https://zenodo.org/records/13885908/files/Chowell_test_Response.tsv
>    ```
>>   <tip-title>Data Type</tip-title>
>>   Leave the `Type` field as `Auto-Detect` when uploading (it's the default).
>>
>    {: .tip}
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>

> > 3. Check that the data format assigned for the file is **tsv**.

Reviewer 2: ```suggestion
> 3. Check that the data format assigned for the file is **tabular**.
```
